internal/cmd/gocdk/deploy.go

```diff
@@ -16,50 +16,172 @@ package main
 
 import (
 	"context"
+	"net/http"
+	"net/url"
+	"path/filepath"
+	"strings"
 
+	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/spf13/cobra"
+	"gocloud.dev/gcp"
+	"gocloud.dev/internal/cmd/gocdk/internal/docker"
+	"gocloud.dev/internal/cmd/gocdk/internal/launcher"
 	"golang.org/x/xerrors"
+	"google.golang.org/api/option"
+	cloudrun "google.golang.org/api/run/v1alpha1"
 )
 
 func registerDeployCmd(ctx context.Context, pctx *processContext, rootCmd *cobra.Command) {
+	var dockerImage string
+	var apply bool
 	deployCmd := &cobra.Command{
 		Use:   "deploy BIOME",
 		Short: "TODO Deploy the biome",
 		Long:  "TODO more about deploy",
 		Args:  cobra.ExactArgs(1),
 		RunE: func(_ *cobra.Command, args []string) error {
-			// Precompute snapshot tag and pass into build and launch.
-			moduleRoot, err := pctx.ModuleRoot(ctx)
-			if err != nil {
-				return xerrors.Errorf("gocdk deploy: %w", err)
-			}
-			imageName, err := moduleDockerImageName(moduleRoot)
-			if err != nil {
-				return xerrors.Errorf("gocdk deploy: %w", err)
-			}
-			tag, err := generateTag()
-			if err != nil {
-				return xerrors.Errorf("gocdk deploy: %w", err)
-			}
-			snapshotRef := imageName + ":" + tag
-			buildRefs := []string{
-				imageName + defaultDockerTag,
-				snapshotRef,
-			}
-
-			// Run build, "biome apply", and launch.
-			biome := args[0]
-			if err := build(ctx, pctx, buildRefs); err != nil {
-				return err
-			}
-			if err := biomeApply(ctx, pctx, biome, true); err != nil {
-				return err
-			}
-			if err := launch(ctx, pctx, biome, snapshotRef); err != nil {
-				return err
-			}
-			return nil
+			return deploy(ctx, pctx, args[0], dockerImage, apply)
 		},
 	}
+	deployCmd.Flags().StringVar(&dockerImage, "image", defaultDockerTag, "Docker image to deploy in the form `name[:tag]` OR `:tag`, or empty string to generate a tag")
```

Comment:
Sorry, one suggestion:

```suggestion
	deployCmd.Flags().StringVar(&dockerImage, "image", defaultDockerTag, "Docker image to deploy in the form `name[:tag]` OR `:tag`, or empty string to build a new image")
```

Reply:
Done.